relay-server/src/actors/global_config.rs

```diff
@@ -271,3 +274,37 @@ impl Service for GlobalConfigService {
         });
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use relay_system::{Controller, Service, ShutdownMode};
+    use relay_test::mock_service;
+    use std::sync::Arc;
+    use std::time::Duration;
+
+    use crate::actors::global_config::{Get, GlobalConfigService};
+
+    /// Tests whether the service shuts down gracefully when the signal is
+    /// received, without requesting more global configs.
+    #[tokio::test]
+    async fn test_service_shutdown() {
+        for mode in &[ShutdownMode::Graceful, ShutdownMode::Immediate] {
+            shutdown_service_with_mode(*mode).await;
+        }
+    }
+
+    async fn shutdown_service_with_mode(mode: ShutdownMode) {
+        let (upstream, _) = mock_service("upstream", (), |&mut (), _| {});
+
+        Controller::start(Duration::from_secs(1));
+
+        let service = GlobalConfigService::new(Arc::default(), upstream).start();
+
+        assert!(service.send(Get).await.is_ok());
+
+        Controller::shutdown(mode);
+        tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
I did not manage to get around tokio to avoid an actual sleep and make this test pass. Open to suggestions!

Reply:
Have you tried `tokio::pause` and `tokio::advance`?